docs/en/lfs.md

```diff
@@ -1,205 +1,144 @@
-
--  I now have an LFS aware version of my LuaOTA provisioning system (see `lua_examples/luaOTA`).  This handles all of the incremental compiling and LFS reloads transparently. This is typically integrated into the ESP application.
+-  If you only need to update the Lua components, then you can work over the air (OTA).  For example see my 
+[HTTP_OTA.lua](https://github.com/nodemcu/nodemcu-firmware/tree/dev/lua_examples/lfs/HTTP_OTA.lua), which pulls a new LFS image from a webservice and reloads it into the LFS region.  This only takes seconds and as a result I often use this in preference to UART-attached loading.
 
 -  Another option would be to include the FTP and Telnet modules in the base LFS image and to use telnet and FTP to update your system.  (Given that a 64Kb LFS can store thousands of lines of Lua, doing this isn't much of an issue.)
 
-My current practice is to use a small bootstrap `init.lua` file in SPIFFS to load the `_init` module from LFS, and this does all of the actual application initialisation. My `init.lua`:
-
--  Is really a Lua binary (`.lc`) file renamed to a `.lua` extension.  Using a binary init file avoids loading the Lua compiler.  This works because even though the firmware looks for `init.lua`, the file extension itself is a just a convention; it is treated as a Lua binary if it has the correct Lua binary header.
+My current practice is to use a small bootstrap `init.lua` file in SPIFFS to load the `_init` module from LFS, and this does all of the actual application initialisation.  As there is a few sec delay whilst connecting to the Wifi, this delay also acts as a "just in case" when I am developing, as this is enough to allow me to paste a `file.remove('init.lua')` into the UART if my test applicaiton is stuck into a panic loop, or do a different development path for debugging.
 
--  Includes a 1 sec delay before connecting to the Wifi.  This is a "just in case" when I am developing.  This is enough to allow me to paste a `file.remove'init.lua'` into the UART if I want to do different development paths.
+Under rare circumstances, for example a power fail during the flashing process, the flash can be left in a part-written state following a `flashreload()`.  The Lua RTS start-up sequence will detect this and take the failsafe opton of resetting the LFS to empty, and if this happens then the LFS `_init` function will be unavailable.  Your `init.lua` should therefore not assume that the LFS contains any modules (such as `_init`), and should contain logic to detect if LFS reset has occurred and if necessary reload the LFS again. Calling `node.flashindex("_init")()` directly will result in a panic loop in these circumstances.  Therefore first check that `node.flashindex("_init")` returns a function or protect the call, `pcall(node.flashindex("_init"))`, and decode the error status to validate that initialisation was successful.
 
```

Comment:
This is pretty far afield and NOT a blocker for the upcoming release, but I wanted to get the thought in your head: any possibility of a dual-slot LFS variant, where `node.flashreload()` can be told which slot to pave over and `node.flashindex()` either always uses the newest or always searches both?  This would give a pretty nice "application LFS" vs. "failsafe LFS" setup and permit an OTA dance where only one of the two was ever potentially invalid at any moment.

Reply:
Ah ha.  So, I have just updated my LFS-flashing bit in `init.lua` to be

```lua
if node.flashindex() == nil then
  -- no LFS image.  Perhaps it failed to flash?
  if file.exists("luac.out.stage") and not file.exists("luac.out") then
    -- Looks like we tried once before, indeed.  Try again.
    node.flashreload("luac.out.stage")
  end
end
file.remove("luac.out.stage") -- remove stage file from last attempt
-- Do we have a new LFS blob?  If so, install it.
if file.exists("luac.out") then
  print("Updating LFS image.  Will reboot if things go well.")
  file.rename("luac.out", "luac.out.stage")
  node.flashreload("luac.out.stage")
  error("Failed to update LFS!")
end
```

If that seems sensible, it's perhaps worth including as an example in the docs?

Reply:
I would rather see problems in the content of the LFS image than in the flashing process. So there could be an early decision from which partition to load if used as alternatives or only to start the base LFS and not the app-LFS part.
This decisoin could be made based on the last run (or its failure) based on bootreason or some other startup markers (a file or RTC mem)

Reply:
> I would rather see problems in the content of the LFS image than in the flashing process. So there could be an early decision from which partition to load if used as alternatives or only to start the base LFS and not the app-LFS part.

The last two full words of an [RFC 1951](https://www.ietf.org/rfc/rfc1951.txt) deflate stream are the CRC and the unpacked len of the output stream.  Having done the first pass which involves reading the entire SPIFFS file, and validated the CRC and the length, we can have confidence that we can read and unpack the image. As I said, the only material risk of failure during reflash is one on power-fail and this will be detected on restart, so whilst having two flash regions seems comfortable, it is just a lot easier to keep any LFS variants that you need as LFS image files on SPIFFS.  As I said, the reflash takes about a second or two, though there is a further few seconds delay in the ROM restart routine.

 Can you give me a scenario where having hot swap LFS regions would add any value at all?

Reply:
@TerryE Terry I did not mean any failure in the flashing process but false Lua code that fails to run for some reason.
I have a module setting a persistent flag at boot time and resetting it after some time.
If at boot it has not been reset (due to application failure) I would like to start a fallback or minimal system which is stored in a safe location (which is never destroyed by a faulty application LFS).

Reply:
Surely you can still do this by a little conditional logic in your `init.lua` and if necessary loading the fallback from SPIFFS?